rest/src/main/java/org/envirocar/server/rest/util/OSMTileRenderer.java

```diff
@@ -157,14 +195,192 @@ BoundingBox tile2boundingBox(final int x, final int y, final int zoom) {
 		return bb;
 	}
 
-	 double tile2lon(int x, int z) {
+	double tile2lon(int x, int z) {
 		return x / Math.pow(2.0, z) * 360.0 - 180;
 	}
 
-	 double tile2lat(int y, int z) {
+	double tile2lat(int y, int z) {
 		double n = Math.PI - (2.0 * Math.PI * y) / Math.pow(2.0, z);
 		return Math.toDegrees(Math.atan(Math.sinh(n)));
 	}
+
+	public BoundingBox findBoundingBoxForGivenLocations(
+			ArrayList<Coordinate> coordinates) {
+		double west = 0.0;
+		double east = 0.0;
+		double north = 0.0;
+		double south = 0.0;
+
+		for (int lc = 0; lc < coordinates.size(); lc++) {
+			Coordinate loc = coordinates.get(lc);
+			// System.out.println(loc.y);
+			if (lc == 0) {
+				north = loc.y;
+				south = loc.y;
+				west = loc.x;
+				east = loc.x;
+
+			} else {
+				if (loc.y > north) {
+					north = loc.y;
+				} else if (loc.y < south) {
+					south = loc.y;
+				}
+				if (loc.x < west) {
+					west = loc.x;
+				} else if (loc.x > east) {
+					east = loc.x;
+				}
+			}
+		}
+		/*
+		 * double padding = 0.001; north = north + padding; south = south -
+		 * padding; west = west - padding; east = east + padding;
+		 */
+		return new BoundingBox(north, south, west, east);
+	}
+
+	public Graphics2D appendImage(BufferedImage newTile, int highestX,
+			int lowestX, int highestY, int lowestY, int zoom)
+			throws IOException {
+		Graphics2D g2d = newTile.createGraphics();
+		System.out.println(lowestX + " : " + lowestY + " : " + highestX + " : "
+				+ highestY);
+		for (int i = lowestX; i <= highestX; i++) {
+			for (int j = lowestY; j <= highestY; j++) {
+				BufferedImage image = downloadTile(i, j, zoom);
+				System.out.println(" i :" + (i - lowestX) * 256 + " j : "
+						+ (j - lowestY) * 256);
+				g2d.drawImage(image, (i - lowestX) * 256, (j - lowestY) * 256,
+						null);
+				// g2d.drawImage(image1,256,0,null);
+				// temp.add(0,256,image.createGraphics());
+				// temp=composite image
+			}
+		}
+		return g2d;
+	}
+
+	public BufferedImage downloadTile(int x, int y, int zoom)
```

Comment:
What about having an internal tile cache? Guava also has a a Cache that someone extended to a file-based cache: http://java.dzone.com/articles/extending-guava-caches May be helpful.

Or this one might be even better: https://github.com/raphw/guava-cache-overflow-extension